src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -634,28 +656,58 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
         token: CancellationToken,
         useCache: 'useCache' | 'ignoreCache' = 'ignoreCache'
     ): Promise<KernelConnectionMetadata[]> {
+        const localKernelConnectionsPromise = this.getLocalKernelConnectionMetadata(
+            listLocalNonPythonKernels,
+            token,
+            useCache
+        );
         if (this.isLocalLaunch) {
-            return listLocalNonPythonKernels
-                ? this.localKernelFinder.listNonPythonKernels(token, useCache)
-                : this.localKernelFinder.listKernels(undefined, token, useCache);
-        } else {
-            if (listLocalNonPythonKernels) {
-                return [];
-            }
-            const ui = new DisplayOptions(false);
-            try {
-                const connection = await this.notebookProvider.connect({
-                    getOnly: false,
-                    resource: undefined,
-                    ui,
-                    localOnly: false,
-                    token
-                });
+            return localKernelConnectionsPromise;
+        }
+        if (listLocalNonPythonKernels) {
+            return [];
+        }
+        const [local, remote] = await Promise.all([
+            localKernelConnectionsPromise,
+            this.getRemoteKernelConnectionMetadata(token)
+        ]);
+        return local.concat(remote);
+    }
 
-                return this.remoteKernelFinder.listKernels(undefined, connection, token);
-            } finally {
-                ui.dispose();
-            }
+    private async getLocalKernelConnectionMetadata(
```

Comment:
To be discussed:
* Currently you can use local & remote together
* I.e. if you have a remote Uri, we will list both (some users have asked for this)

Not sure whether this is a good idea, perhaps we can turn this into a flag/setting when users defines the remote Uri, we can ask them whether they'd like to see the local kernel as well (instead of a setting, else they'll never discover that).